providers/fips/fipsprov.c

```diff
@@ -88,12 +146,84 @@ int OSSL_provider_init(const OSSL_PROVIDER *provider,
         case OSSL_FUNC_CORE_GET_PARAMS:
             c_get_params = OSSL_get_core_get_params(in);
             break;
+        case OSSL_FUNC_CORE_PUT_ERROR:
+            c_put_error = OSSL_get_core_put_error(in);
+            break;
+        case OSSL_FUNC_CORE_ADD_ERROR_VDATA:
+            c_add_error_vdata = OSSL_get_core_add_error_vdata(in);
+            break;
         /* Just ignore anything we don't understand */
         default:
             break;
         }
     }
 
+    if (!ex_data_init_intern())
+        return 0;
+
+    if (!default_context_init_intern()) {
+        crypto_cleanup_all_ex_data_int();
+        return 0;
+    }
+    if (!provider_store_init_intern()) {
+        do_default_context_deinit();
+        crypto_cleanup_all_ex_data_int();
+        return 0;
+    }
+    if (!ossl_method_store_init()) {
+        do_default_context_deinit();
+        crypto_cleanup_all_ex_data_int();
+        return 0;
+    }
+    if (!default_method_store_init()) {
+        fips_teardown();
+        return 0;
+    }
```

Comment:
So....  instead of all of that, could we simply have a provider module global library context, i.e. use OPENSSL_CTX_new() to create it and OPENSSL_CTX_free() to free it, and use it everywhere we can?  It seems like a saner approach than doing all kinds of internal calls, with all the entanglement that could result in in the future.

Reply:
Yes, I think that would be nice. We should use OPENSSL_CTX wherever we have global state we need to initialise. But I see that as outside of the scope of this PR.

Reply:
I don't.  Basically, you've made a load of static initialization functions more widely available, which has a risk of entanglement I already see with current init functionality.  You could achieve the exact same thing by simply creating a library context within the FIPS provider, and by freeing it on teardown.  No need to go call all those internal initializers from outside (or should not be any need at the very least).

Since you do already include context.c in the FIPS module, you already have all you need to do this, so leaving it for another PR will basically mean undoing the bunch of init stuff you've entangled yourself with here.  Therefore, I'd say this is very much *in* scope.